### PR TITLE
Use quaternion rotation map in BinaryCompactObject domain creator

### PR DESCRIPTION
**Pull request by knelli2**
Control systems will require us to use the quaternion rotation map.

### Upgrade instructions

When specifying input options for the `BinaryCompactObject` domain creator, replace
```
RotationAboutZAxisMap:
  InitialRotationAngle: 1.0
  InitialAngularVelocity: -0.5
```
with
```
RotationMap:
  InitialAngularVelocity: [0.0, 0.0, -0.5]
```
Note: The initial rotation angle is now always assumed to be zero and cannot be specified by the user.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp
Comment (wthrowe) [inline src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp]: const args.

File: src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.hpp
Comment (wthrowe) [inline src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.hpp]: Which set of angles is this?  From the cpp I think they're yaw-pitch-roll (permuted?).

File: src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp
Comment (wthrowe) [inline src/Domain/CoordinateMaps/TimeDependent/RotationMatrixHelpers.cpp]: You've got issues with branch cuts here (the correct values can be negative), and switching equations based on roundoff differences usually indicates bad numerical behavior.

Wikipedia [suggests an algorithm](https://en.wikipedia.org/wiki/Rotation_matrix#Quaternion).  I haven't checked it, but it looks sane.

Add tests that would have caught this.

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (wthrowe) [inline src/Domain/Creators/BinaryCompactObject.hpp]: Update.

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (wthrowe) [inline tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp]: Can we split the slow part of this up?

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (wthrowe) [inline src/Domain/Creators/BinaryCompactObject.hpp:466]: If I'm reading this all correctly, the input file expects something like Euler angles for the initial angle, but uses an axis-angle formulation for the initial velocity.  Could we do axis-angle for both?  (Trying to do velocities with Euler angles sounds like a mess.)

**Review by wthrowe — CHANGES_REQUESTED** [review]
For the upgrade instructions: what if the values aren't all zero?

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (knelli2) [reply src/Domain/Creators/BinaryCompactObject.hpp:466]: Yes I think your suggestion of using axis-angle for both is best. Then it's trivial to get the initial quaternion from inputs. Doing that fixes several of the other issues you raised here.

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (knelli2) [reply tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp]: The "slow" parts of this test are the `test_bbh_time_dependent_factory()` tests which take up about ~90% of this test with these changes. The reason for this is the `QuaternionFunctionOfTime`. Every time it is called, it solves an ODE to get the quaternion, integrating from the last update time to the time requested. However, the "last update time" here is just the initial time. So every time we want the quaternion we are integrating from 0.0 to 7.8 (the last tested time for this test, just as an example) with a tolerance of 1.0e-12. Integrating this far only a couple times is still pretty quick, but I calculated that there are ~83,000 time-dep coord map evaluations in this whole test...so things add up and it gets pretty slow.

One thing I thought about doing was changing the times to check from `{0.0, 4.4, 7.8}` to something like `{0.0, 0.4, 1.1, 1.6}`. This should keep the test under 15 seconds. I was just worried about restricting the times tested to small times.

**Comment by knelli2** [comment]
Added a fixup and reverted one commit because I didn't need it anymore on account of using only the axis-angle representation.
Edit: Also fixed PR description

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (wthrowe) [inline src/Domain/Creators/BinaryCompactObject.hpp]: Why can the initial rotation only be about the angular velocity?

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (wthrowe) [reply tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp]: Yeah, I think that's better.  When the test tolerances are ~1e-10 and the timescales are ~0.2, a time of 1.6 isn't really very small.

**Review by wthrowe — COMMENTED** [review]
You can squash the existing fixup.

File: src/Domain/Creators/BinaryCompactObject.hpp
Comment (knelli2) [reply src/Domain/Creators/BinaryCompactObject.hpp]: When using the axis-angle representation, the angle is about the axis. The axis here *is* the angular velocity, so the angle is about the angular velocity.

Comment (wthrowe) [reply src/Domain/Creators/BinaryCompactObject.hpp]: But why do the value and derivative have to share the same axis?

Comment (knelli2) [reply src/Domain/Creators/BinaryCompactObject.hpp]: Now I'm questioning if we even need a rotation angle. There are a couple notions of "angle" here and I want to make sure we are talking about the same ones. A unit quaternion represented as q = (cos(theta/2), nhat*sin(theta/2)) will rotate a vector about nhat by an angle theta. Here, nhat is the angular velocity. So "theta" is one notion of angle about the angular velocity. Then there are phi_x, phi_y, phi_z where the x-component of the angular velocity omega_x = dt(phi_x), omega_y = dt(phi_y), etc. This is another notion. I was referring to "theta" in that description, but you seem to be referring to "phi_x" etc. when you say value and derivative.

For initial conditions, we must specify the initial quaternion and initial angular velocity to properly construct things. But since the vector part of the quaternion is just the unit vector in the direction of the angular velocity, we technically only need "theta" to specify the quaternion. After thinking about it, I think we can just set "theta", "phi_x", etc. = 0 always for initial conditions, so the initial quaternion is just (1,0,0,0). (For reference, for SpEC ID, you only specify omega_z and the quaternion is assumed to be (1,0,0,0), implying that theta = 0)

Also, for our initial conditions, are we ever going to start a binary where the angular velocity isn't initially aligned with the z-axis? If we are, then I can leave the InitialAngularVelocity as a vector. But if not, I think we can make it a scalar again and have it just be the z-component of the initial vector.

Thoughts?

Comment (wthrowe) [reply src/Domain/Creators/BinaryCompactObject.hpp]: I don't think there's a reason to have more freedom here than in SpEC.  Maybe there are cases where a slightly off-axis initial rotation would be better, but I haven't heard about anyone having problems with the SpEC version so any such effect is probably small enough that the control system can quickly compensate.

**Comment by knelli2** [comment]
Rebased, squashed in previous fixup, added new fixup.

@wthrowe I decided to remove the rotation angle input option and leave the initial angular velocity as a vector.

**Comment by wthrowe** [comment]
Looks fine.  Go ahead and squash again.

**Review by kidder — DISMISSED** [review]
This LGTM.  @wthrowe this has been squashed.  @geoffrey4444 are you planning on looking at this, or can we merge it?

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (nilsvu) [inline tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:811]: Is it the quaternion ODE integration that slows this down? Is that fine?

Comment (nilsvu) [reply tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:811]: I saw your explanation above. So the answer is that in an evolution the ODE integration only happens occasionally and hence doesn't slow things down a lot, but in this test it's called an excessive number of times?

**Review by wthrowe — DISMISSED** [review]
Code looks fine, but upgrade instructions need to be updated.

Not merging yet to wait for a response to @nilsvu's question.

File: tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp
Comment (knelli2) [reply tests/Unit/Domain/Creators/Test_BinaryCompactObject.cpp:811]: Correct. The quaternion ODE integration is comparable in time to the one in SpEC.

**Comment by knelli2** [comment]
Updated the upgrade instructions.

**Comment by nilsvu** [comment]
@knelli2 looks like you have to increase the timeout further or reduce the number of ODE integrations for the test to pass. gcc-8 Debug is failing.

**Comment by knelli2** [comment]
Rebased and decreased number of ODE integrations by reducing the number of times checked in the test (the range is still the same, just less intermediate times tested).

**Comment by kidder** [comment]
ping @geoffrey4444

**Comment by knelli2** [comment]
Fixed conflicts from #3857

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
I am going to merge this later today unless I hear that anyone has any remaining issues @wthrowe @nilsvu @geoffrey4444

**Pull request merged**
